# Main entrypoint of the Pseudobatos workflow.
# @authors: Luis Kress, Johannes Hausmann
# Please follow the best practices: 
# https://snakemake.readthedocs.io/en/stable/snakefiles/best_practices.html,
# in particular regarding the standardized folder structure mentioned there. 

import pandas as pd
import os

print(workflow.source_path("../config/config.yml"))

configfile: workflow.source_path("../config/config.yml")

wildcards = glob_wildcards("data/{dataset}/{gene}.fasta")


rule all:
    input:
        expand('results/{dataset}/trees/SUPERTREE.nwk', dataset=wildcards.dataset)

rule mafft:
    input:
        fasta = 'data/{dataset}/{gene}.fasta'
    output:
        aligned_fa = 'results/{dataset}/msa/{gene}.aligned.fasta'
    params:
        maxiterate = config['mafft']['maxiterate']
    threads:
        config['mafft']['threads']
    shell:
        '''
        mafft --maxiterate {params.maxiterate} --globalpair --thread {threads} {input.fasta} > {output.aligned_fa}
        '''


rule trimming:
    input:
        alignment = rules.mafft.output
    output:
        trimmed = 'results/{dataset}/msa/{gene}.trimmed.fasta'
    conda:
        'envs/clipkit.yml'
    shell:
        '''
	clipkit {input.alignment} -m kpic-smart-gap -o {output.trimmed}
        '''

rule supertree:
    input:
        trimmed = rules.trimming.output.trimmed
    output:
        tree = 'results/{dataset}/msa/{gene}.trimmed.fasta.treefile'
    threads:
        config['iqtree']['threads']
    shell:
        '''
        /opt/software/iqtree-2.2.0/bin/iqtree2 -s {input.trimmed} -redo -bb 1000 -nt 2
        '''

rule concat_trees:
    input:
        trees = expand('results/{dataset}/msa/{gene}.trimmed.fasta.treefile', zip, gene=wildcards.gene, dataset=wildcards.dataset)
    output:
        genetrees = 'results/{dataset}/trees/GENETREES.nwk'
    shell:
        '''
        cat {input.trees} > {output.genetrees}
        '''

rule astral:
    input:
        genetrees = rules.concat_trees.output.genetrees
    output:
        supertree = 'results/{dataset}/trees/SUPERTREE.nwk'
    shell:
        '''
        java -jar /home/student7/phylogeny/data/mitochondrion_proteins/Astral/astral.5.7.3.jar -i {input.genetrees} -o {output.supertree}
        '''

